functionalDB/chapter.md

```diff
@@ -0,0 +1,1098 @@
+         relevant-attrs (filter #((usage-pred index) %) all-attrs)
+         add-in-index-fn (fn [ind attr] 
+                                 (update-attr-in-index ind ent-id (:name attr) (:value attr) :db/add))]
+        (assoc layer ind-name  (reduce add-in-index-fn index relevant-attrs))))
+
+(defn- update-attr-in-index [index ent-id attr-name target-val operation]
+   (let [colled-target-val (collify target-val)
+         update-entry-fn (fn [indx vl] 
+                                 (update-entry-in-index indx ((from-eav index) ent-id attr-name vl) operation))]
+     (reduce update-entry-fn index colled-target-val)))
+     
+(defn- update-entry-in-index [index path operation]
+   (let [update-path (butlast path)
+         update-value (last path)
+         to-be-updated-set (get-in index update-path #{})]
+     (assoc-in index update-path (conj to-be-updated-set update-value))))
+````
+All that work is now added as a new layer to the database that we started with, and all that’s left is to update the database’s timestamp and top-id fields to prepare ourselves for future interactions. That last step is done at the last line of *add-entity*, that also returns the updated database.
+
+Note that it is possible to add several entities in one function call, simply by calling the *add-entities* functions, that all it does is going over the given collection of entities and add them one by one to the database.
+
+````clojure
+(defn add-entities [db ents-seq] (reduce add-entity db ents-seq))
+````
+#### Removing an entity
+
+When an entity is removed from the database, it means that a new layer is constructed, one that does not have any trace of that entity. 
+
+When removing an entity’s trace, we do not need only to remove the entity, but also affect other entities that refer to it, as well as clear that referencing from the indexing system.
+
+This entire cleanup process (or more correctly - a "construct without" process) is managed by the *remove-entity* function. In order not to tire the reader, we’ll skip the part where the entity itself is removed (this part is quite similar to the addition process described before), and we’ll focus on the reference removal part.
+
+````clojure
+(defn remove-entity [db ent-id]
+   (let [ent (entity-at db ent-id)
+         layer (remove-back-refs db ent-id (last (:layers db)))
+         no-ref-layer (update-in layer [:VAET] dissoc ent-id)
+         no-ent-layer (assoc no-ref-layer :storage 
+                                   (drop-entity  
+                                          (:storage no-ref-layer) ent))
+         new-layer (reduce (partial remove-entity-from-index ent) 
+                                 no-ent-layer (indices))]
+     (assoc db :layers (conj  (:layers db) new-layer))))
+````
+The reference removal part is handled in the *remove-back-refs* function.
+
+````clojure
+(defn- remove-back-refs [db e-id layer]
+   (let [refing-datoms (reffing-datoms-to e-id layer)

```

Comment:
This is a nitpick, but why do we use 1 f in the function name `refing-datoms`, and then two fs in `reffing-datoms`? :)